src/key_value_store.rs

```diff
@@ -113,3 +113,148 @@ impl<
+}
+
+impl<I, C: Eq> Persistable<I, C> {
+    /// Commit the pending changes.
+    pub fn commit(&mut self) {
+        // The index is always up-to-date, so commting only needs to clear the pending changes.
+        self.pending_changes = Vec::new();
+    }
+}
+
+pub type PersistableHashSet<K> = Persistable<HashSet<K>, K>;
+pub type PersistableHashMap<K, V> = Persistable<HashMap<K, V>, (K, V)>;
+pub type PersistableBTreeMap<K, V> = Persistable<BTreeMap<K, HashSet<V>>, (K, V)>;
+
+impl<K: Copy + Eq + Hash> Persist<K> for PersistableHashSet<K> {
+    fn new() -> Self {
+        Self {
+            index: HashSet::new(),
+            pending_changes: Vec::new(),
+        }
+    }
+
+    fn insert(&mut self, change: K) {
+        self.index.insert(change);
+        self.pending_changes.push(IndexChange::Add(change));
+    }
+
+    fn remove(&mut self, change: K) {
+        self.index.remove(&change);
+        self.pending_changes.push(IndexChange::Remove(change));
+    }
+
+    fn revert(&mut self) {
+        for change in &self.pending_changes {
+            match change {
+                IndexChange::Add(key) => {
+                    self.index.remove(key);
+                }
+                IndexChange::Remove(key) => {
+                    self.index.insert(*key);
+                }
+            }
+        }
+        self.pending_changes = Vec::new();
+    }
+}
+
+impl<K: Clone + Eq + Hash, V: Clone + Eq + Hash> Persist<(K, V)> for PersistableHashMap<K, V> {
+    fn new() -> Self {
+        Self {
+            index: HashMap::new(),
+            pending_changes: Vec::new(),
+        }
+    }
+
+    fn insert(&mut self, change: (K, V)) {
+        self.index.insert(change.0.clone(), change.1.clone());
+        self.pending_changes.push(IndexChange::Add(change));
+    }
+
+    fn remove(&mut self, change: (K, V)) {
+        self.index.remove(&change.0);
+        self.pending_changes.push(IndexChange::Remove(change));
+    }
+
+    fn revert(&mut self) {
+        for change in &self.pending_changes {
+            match change {
+                IndexChange::Add((key, _)) => {
+                    self.index.remove(key);
+                }
+                IndexChange::Remove((key, value)) => {
+                    self.index.insert(key.clone(), value.clone());
+                }
+            }
+        }
+        self.pending_changes = Vec::new();
+    }
+}
+
+impl<K: Copy + Eq + Hash + Ord, V: Clone + Eq + Hash> Persist<(K, V)>
+    for PersistableBTreeMap<K, V>
+{
+    fn new() -> Self {
+        Self {
+            index: BTreeMap::new(),
+            pending_changes: Vec::new(),
+        }
+    }
+
+    fn insert(&mut self, change: (K, V)) {
+        self.index
+            .entry(change.0)
+            .or_insert_with(HashSet::new)
+            .insert(change.1.clone());
+        self.pending_changes.push(IndexChange::Add(change));
+    }
+
+    fn remove(&mut self, change: (K, V)) {
+        self.index.remove(&change.0);
```

Comment:
Doesn't this remove _all_ values with key `K`, when we only want to remove the value `V`?

Reply:
Overall I'm kind of wondering if this is the right interface for this data structure?

Reply:
Yeah seems reasonable (one more comment on the implementation but this seems like the correct functionality for this interface to have). We may eventually want to implement more methods for `PersistableBTreeMultiMap` than just the `Persist` trait, since it's more complicated than a simple key-value map. Like `remove_all` for example. But we can just see what needs arise.